protocol/synthetix/contracts/storage/AccountDelegationIntents.sol

```diff
@@ -0,0 +1,170 @@
+//SPDX-License-Identifier: MIT
+pragma solidity >=0.8.11 <0.9.0;
+
+import "./DelegationIntent.sol";
+import "@synthetixio/core-contracts/contracts/utils/SetUtil.sol";
+import "@synthetixio/core-contracts/contracts/utils/SafeCast.sol";
+import "../interfaces/IVaultModule.sol";
+import "./Account.sol";
+
+/**
+ * @title Represents a delegation (or undelegation) intent.
+ */
+library AccountDelegationIntents {
+    using SafeCastI256 for int256;
+    using SafeCastU128 for uint128;
+    using SafeCastU256 for uint256;
+    using SetUtil for SetUtil.UintSet;
+    using SetUtil for SetUtil.AddressSet;
+    using DelegationIntent for DelegationIntent.Data;
+    using Account for Account.Data;
+
+    struct Data {
+        uint128 accountId;
+        uint128 delegationIntentsEpoch; // nonce used to nuke previous intents using a new era (useful on liquidations)
+        SetUtil.UintSet intentsId;
+        mapping(bytes32 => SetUtil.UintSet) intentsByPair; // poolId/collateralType => intentIds[]
```

Comment:
can we just embed `DelegationIntent.Data[]` type right here? then we:
1. don't need to store `intentsId`
2. don't need to store `id` inside of the `DelegationIntent`
3. since intents are added in timestamp order, we can wipe out expired intents very cleanly without searching the full intents list by creating another field `lastActiveIntent` and incrementing that.
4. if we make it a rule that intents have to be processed in the order they are submitted, this same field can be used to invalidate intents as we go

also if an intent is "cancelled" by the owner before it can be processed, we can add a field to the intent "cancelled, which only incurs a cost on intent processing and nowhere else

Reply:
All in favor optimizing the removal of id @dbeal-eth   and replacing with timestamps, in case the lp/account-id/market is the key to holding the stored intents... In order to avoid different timestamps stepping over each other. The the rule to make intents processed in the order they are submitted can lead to issues with vaults... Assume the delay / window are configured differently for delegate / undelegate and a user specifies a undelegation on a vault and another user a delegate.. They can lead to stepping over each other... So this would require 2 intent data's right? One for delegation and another for undelegation?
Also another scenario I was thinking about is assume you have an intent that can't be expired or executed, and is in limbo until expiry, Example caps. This leads to kind of bricking until parameters are updated... So enforcing sequentiality I thought about this and opted to not ask for it to be employed purposefully....

Reply:
yea 3 is literally intended tobe a simplification for clients by allowing for the contract to be able to handle everything effieicntly.

regarding #4, if delegate/undelegate have different window or delay, then we can separate the lists by "operation" (its already separated by vault). Tho to simplify things I dont see a lot of use in allowing different window limits for delegate/undelegate...

regarding update of parameters... I mean these settings apply immediately to all pending intents regardless of when they are created and seems like it would be an issue with either scheme, so seems like a small issue if anything comparing to the benefits of the optimizations.

Reply:
There is also the situation of an intent that is neither executable and the window is not out for it's expiry, blocking the vault from executing any other intent. Aside from parameter changes, there are things that could go wrong and requiring sequential execution for async intents I feel is too fragile for the sake of optimization.